### PR TITLE
CORTX-31244: CSM session is not getting removed from consul after expiry time

### PR DESCRIPTION
**Pull request by Pranali04796**
Signed-off-by: Pranali Ugale <pranali.ugale@seagate.com>

# Pull Request
## Problem Statement
-   [CORTX-31244](https://jts.seagate.com/browse/CORTX-31244)

## Design
-   Write background process to remove sessions after periodic amount of time

-   \[ \] Coding conventions are followed and code is consistent

-   \[ \] Unit and System Tests are added
-   \[ \] Test Cases cover Happy Path, Non-Happy Path and Scalability
-   \[ \] Testing was performed with RPM
- o1)
Set expiry limit to 30 seconds. Set background process to delete expired token in 1 minute.
- o2

-   \[ \] Interface change (if any) are documented
-   \[ \] Side effects on other features (deployment/upgrade)
-   \[ \] Dependencies on other component(s)

-   \[ \] JIRA number/GitHub Issue added to PR
-   \[ \] PR is self reviewed
-   \[ \] Jira and state/status is updated and JIRA is updated with PR link
-   \[ \] Check if the description is clear and explained

-   \[ \] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/core/agent/api.py [hunk 1]
```diff
@@ -592,6 +593,16 @@ async def _ssl_cert_check_bg(cls):
 
         Log.debug('SSL certificate expiry check background task done')
 
+    @classmethod
+    async def _clear_expired_sessions_bg(cls):
+        Log.debug('Started background task for clearing expired sessions')
```
Comment (pawankumarsrivastava) [inline csm/core/agent/api.py:598]: Change it to info

File: csm/core/agent/api.py [hunk 2]
```diff
@@ -592,6 +593,16 @@ async def _ssl_cert_check_bg(cls):
...
+        try:
+            session_mgr_service = cls._app[const.SESSION_MGR_SERVICE]
+            await session_mgr_service.clear_sessions()
+        except AsyncioCancelledError:
+            Log.error('Background task for clearing expired session cancelled')
+        Log.debug('Background task for clearing expired session done')
```
Comment (pawankumarsrivastava) [inline csm/core/agent/api.py:604]: change it to info

File: csm/core/services/session/session_factory.py [hunk 3]
```diff
@@ -97,7 +98,9 @@ def get_user_role(self) -> Optional[str]:
         creds = self._credentials
         return creds.user_role if isinstance(creds, LocalCredentials) else None
 
-
+    def is_expired(self) -> bool:
+        """Check if the session is expired."""
+        return datetime.now(timezone.utc) > self._expiry_time
```
Comment (pawankumarsrivastava) [inline csm/core/services/session/session_factory.py:103]: Add a space between the functions

File: csm/core/services/sessions.py [hunk 4]
```diff
@@ -101,6 +102,34 @@ async def get_all(self):
...
+        """
+        Entry point for clearing expired token background task
+        :return:
+        """
+        today= datetime.now(timezone.utc).date()
+        await self._timer_task(handler=self._remove_expired_sessions,
+        start=datetime(today.year, today.month, today.day,tzinfo=timezone.utc),interval=timedelta(seconds=2))
+
+    async def _timer_task(self, handler, start: datetime, interval: timedelta):
+        current = datetime.now(timezone.utc)
```
Comment (pawankumarsrivastava) [inline csm/core/services/sessions.py:115]: No doc strings for functions

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/core/agent/api.py (hunk 1 above)
Comment (Pranali04796) [reply csm/core/agent/api.py]: resolved

File: csm/core/services/session/session_factory.py (hunk 3 above)
Comment (Pranali04796) [reply csm/core/services/session/session_factory.py:103]: Added

File: csm/core/services/sessions.py (hunk 4 above)
Comment (Pranali04796) [reply csm/core/services/sessions.py:119]: resolved

File: csm/core/agent/csm_agent.py [hunk 5]
```diff
@@ -104,6 +104,7 @@ def init():
         user_manager = UserManager(db)
         role_manager = RoleManager(roles)
         session_manager = SessionManager(db)
+        CsmRestApi._app["session_manager"] = session_manager
```
Comment (rohitkolapkar) [inline csm/core/agent/csm_agent.py:107]: use  CsmRestApi._app[const.SESSION_MGR_SERVICE ]

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

**Review by rohitkolapkar — APPROVED** [review]

**Pull request merged**